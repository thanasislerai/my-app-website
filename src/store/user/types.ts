import firebase from "firebase/app";

export interface User {
  id: string;
  userName: string;
  email: firebase.User["email"];
  token?: string;
  imageUrl?: string;
}

export interface UserState {
  userInfo?: User;
  loading: boolean;
  error?: string | null;
  firebaseUserLoading?: boolean;
}

export interface UserSignInParams {
  email: string;
  password: string;
}

export interface UserSignUpParams {
  userName: string;
  email: string;
  password: string;
  imageUrl?: string;
}

export interface UserResponse {
  _id: string;
  userName: string;
  email: string;
  imageUrl?: string;
}
