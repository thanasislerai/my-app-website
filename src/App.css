html,
body {
  height: 100%;
  min-height: 100%;
}

#root {
  display: flex;
  flex-direction: column;
  height: 100%;
}

.mapboxgl-ctrl-bottom-left, .mapboxgl-ctrl-bottom-right {
  display: none;
}
