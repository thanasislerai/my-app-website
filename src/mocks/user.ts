import { User } from "../store/user/types";

export const mockUser: User = {
  id: "RaNdOmStRiNg",
  email: "test@mail.com",
  token: "SoMeRaNdOmStRiNg",
  userName: "testuser",
  firebaseUid: "SoMeRaNdOmStRiNg",
};
