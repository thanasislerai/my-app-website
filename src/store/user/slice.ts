import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";
import { AxiosError } from "axios";
import userServices from "../../services/userServices";
import type { CreateAsyncThunkTypes } from "../configure";

import { UserSignInParams, UserSignUpParams, UserState } from "./types";

const userInitialState: UserState = {
  loading: false,
};

export const signInUser = createAsyncThunk<
  UserState["userInfo"],
  UserSignInParams,
  CreateAsyncThunkTypes
>("user/signIn", async ({ email, password }, { rejectWithValue }) => {
  try {
    const { user } = (await userServices.signIn(email, password)) || {};
    const token = await user?.getIdToken();
    const { data: storedUser } = await userServices.getSelf(token);

    return {
      userName: storedUser.userName,
      email: storedUser.email,
      token,
    };
  } catch (err: any) {
    const error: AxiosError<UserState["error"]> = err;
    return rejectWithValue(error.message);
  }
});

export const signUpUser = createAsyncThunk<
  UserState["userInfo"],
  UserSignUpParams,
  CreateAsyncThunkTypes
>("user/signUp", async ({ email, password, userName }, { rejectWithValue }) => {
  try {
    const { user: firebaseUser } =
      (await userServices.signUp(email, password)) || {};
    const token = await firebaseUser?.getIdToken();
    const { data: storedUser } = await userServices.storeUser(userName, email);

    return {
      userName: storedUser.userName,
      email: storedUser.email,
      token,
    };
  } catch (err: any) {
    const error: AxiosError<UserState["error"]> = err;
    return rejectWithValue(error.message);
  }
});

export const getSelf = createAsyncThunk<
  UserState["userInfo"],
  string,
  CreateAsyncThunkTypes
>(
  "user/getSelf",
  async (token: string, { rejectWithValue }) => {
    try {
      const { data: userData } = await userServices.getSelf(token);

      return {
        email: userData.email,
        userName: userData.userName,
        token,
      };
    } catch (err: any) {
      const error: AxiosError<UserState["error"]> = err;
      return rejectWithValue(error.message);
    }
  },
  {
    // If another user action is pending, cancel this request before it starts.
    condition(arg: string, { getState }) {
      const {
        user: { loading },
      } = getState();
      return !loading;
    },
  }
);

export const signOutUser = createAsyncThunk<
  UserState["userInfo"],
  undefined,
  CreateAsyncThunkTypes
>("user/signOut", async (_, { rejectWithValue }) => {
  try {
    await userServices.signOut();
    return undefined;
  } catch (err: any) {
    const error: AxiosError<UserState["error"]> = err;
    return rejectWithValue(error.message);
  }
});

const userSlice = createSlice({
  name: "user",
  initialState: userInitialState,
  reducers: {
    setUser: (state, action: PayloadAction<UserState["userInfo"]>) => ({
      ...state,
      userInfo: action.payload,
    }),
    clearError: (state) => ({ ...state, error: undefined }),
  },
  extraReducers: (builder) => {
    // User sign in
    builder.addCase(
      signInUser.fulfilled,
      (state, action: PayloadAction<UserState["userInfo"]>) => ({
        ...state,
        userInfo: action.payload,
        loading: false,
      })
    );

    builder.addCase(signInUser.pending, (state) => ({
      ...state,
      loading: true,
      error: null,
    }));

    builder.addCase(
      signInUser.rejected,
      (state, action: PayloadAction<UserState["error"]>) => {
        return {
          ...state,
          error: action.payload,
          loading: false,
        };
      }
    );

    // User sign out
    builder.addCase(
      signOutUser.fulfilled,
      (state, action: PayloadAction<UserState["userInfo"]>) => ({
        ...state,
        userInfo: action.payload,
        loading: false,
      })
    );

    builder.addCase(signOutUser.pending, (state) => ({
      ...state,
      loading: true,
      error: null,
    }));

    builder.addCase(
      signOutUser.rejected,
      (state, action: PayloadAction<UserState["error"]>) => {
        return {
          ...state,
          error: action.payload,
          loading: false,
        };
      }
    );

    // User sign up
    builder.addCase(
      signUpUser.fulfilled,
      (state, action: PayloadAction<UserState["userInfo"]>) => ({
        ...state,
        userInfo: action.payload,
        loading: false,
      })
    );

    builder.addCase(signUpUser.pending, (state) => ({
      ...state,
      loading: true,
      error: null,
    }));

    builder.addCase(
      signUpUser.rejected,
      (state, action: PayloadAction<UserState["error"]>) => {
        return {
          ...state,
          error: action.payload,
          loading: false,
        };
      }
    );

    // Get self
    builder.addCase(
      getSelf.fulfilled,
      (state, action: PayloadAction<UserState["userInfo"]>) => ({
        ...state,
        userInfo: action.payload,
        loading: false,
      })
    );

    builder.addCase(getSelf.pending, (state) => ({
      ...state,
      loading: true,
      error: null,
    }));

    builder.addCase(
      getSelf.rejected,
      (state, action: PayloadAction<UserState["error"]>) => {
        return {
          ...state,
          error: action.payload,
          loading: false,
        };
      }
    );
  },
});

export const { setUser, clearError } = userSlice.actions;

export default userSlice.reducer;
