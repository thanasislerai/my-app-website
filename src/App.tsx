import React from "react";
import "./App.css";
import { ThemeProvider } from "@material-ui/core";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";

import theme from "./theme";
import NavBar from "./common/NavBar";
import HomePage from "./routes/HomePage";
import NotFound from "./common/NotFound";

function App() {
  return (
    <ThemeProvider theme={theme}>
      <NavBar />
      <Router>
        <Switch>
          <Route exact path="/" component={HomePage} />
          <Route component={NotFound} />
        </Switch>
      </Router>
    </ThemeProvider>
  );
}

export default App;
