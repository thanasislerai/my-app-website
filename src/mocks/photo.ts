import { Photo } from "../store/user/types";

export const mockPhoto: Photo = {
  id: "2",
  isPublic: true,
  lat: 51.50861,
  lng: -0.163611,
  url: "https://c4.wallpaperflare.com/wallpaper/344/391/747/hyde-park-in-autumn-london-england-wallpaper-preview.jpg",
  userId: "1",
  title: "First photo of Hyde Park",
};
