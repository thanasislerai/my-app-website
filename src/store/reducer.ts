import { combineReducers } from "redux";

import theme from "./theme/slice";
import user from "./user/slice";

const appReducer = combineReducers({
  theme,
  user,
});

export default appReducer;
