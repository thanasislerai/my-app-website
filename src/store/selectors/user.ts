import { RootState } from "../configure";
import { UserState } from "../user/types";

export const userInfoSelector = (state: RootState): UserState["userInfo"] =>
  state.user.userInfo;

export const userLoadingSelector = (state: RootState): UserState["loading"] =>
  state.user.loading;

export const userErrorSelector = (state: RootState): UserState["error"] =>
  state.user.error;

export const firebaseUserLoadingSelector = (
  state: RootState
): UserState["firebaseUserLoading"] => state.user.firebaseUserLoading;
