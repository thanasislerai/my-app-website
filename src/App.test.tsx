import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";

import App from "./App";
import { mockUser } from "./mocks/user";
import { RootState } from "./store/configure";

const mockStore = configureStore<RootState>();

jest.mock("./routes/HomePage/Map", () => "MockMap");

describe("App", () => {
  let element: HTMLElement;
  beforeEach(() => {
    const store = mockStore({
      theme: {
        type: "dark",
      },
      user: {
        userInfo: mockUser,
        loading: false,
        error: null,
      },
    });

    store.dispatch = jest.fn();

    element = render(
      <Provider store={store}>
        <App />
      </Provider>
    ).container;
  });

  it("renders as expected", () => {
    expect(element).toMatchSnapshot();
  });
});
