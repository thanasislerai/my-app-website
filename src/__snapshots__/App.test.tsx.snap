// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders as expected 1`] = `
<div>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <div
        class="PrivateSwipeArea-root-14 PrivateSwipeArea-anchorLeft-15"
        style="width: 20px;"
      />
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-space-between"
      >
        <div
          class="MuiGrid-root MuiGrid-item"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item"
        >
          <div
            class="MuiGrid-root makeStyles-buttonGroup-1 MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <a
                aria-disabled="false"
                class="MuiButtonBase-root MuiIconButton-root"
                href="/"
                role="button"
                tabindex="0"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                tabindex="0"
                title="Switch to dark theme"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M22.6 11.29L20 8.69V5c0-.55-.45-1-1-1h-3.69l-2.6-2.6a.9959.9959 0 00-1.41 0L8.69 4H5c-.55 0-1 .45-1 1v3.69l-2.6 2.6c-.39.39-.39 1.02 0 1.41L4 15.3V19c0 .55.45 1 1 1h3.69l2.6 2.6c.39.39 1.02.39 1.41 0l2.6-2.6H19c.55 0 1-.45 1-1v-3.69l2.6-2.6c.39-.39.39-1.03 0-1.42zm-4.68 1.69c-.34 2.12-1.85 3.94-3.88 4.66-1.21.43-2.41.45-3.5.18-.41-.1-.48-.65-.13-.9C11.98 15.84 13 14.04 13 12s-1.02-3.84-2.58-4.92c-.35-.24-.29-.79.13-.9 1.09-.27 2.29-.25 3.5.18 2.02.72 3.54 2.54 3.88 4.66.05.33.07.66.07.98-.01.32-.03.65-.08.98z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-center"
              >
                <div
                  class="MuiGrid-root MuiGrid-item"
                >
                  <div
                    class="MuiAvatar-root MuiAvatar-circular makeStyles-profileAvatar-10 MuiAvatar-colorDefault"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiAvatar-fallback"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextPrimary"
                  >
                    testuser
                  </p>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </header>
  <div
    class="MuiBox-root MuiBox-root-12"
  >
    <div
      class="MuiBox-root MuiBox-root-13"
    >
      <h2
        class="MuiTypography-root MuiTypography-h2 MuiTypography-colorTextPrimary MuiTypography-gutterBottom MuiTypography-alignCenter"
      >
        Welcome!!
      </h2>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Start Journey
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
